internal/dot/graph.go

```diff
@@ -130,21 +163,84 @@ func (dg *Graph) AddCtor(c *Ctor, paramList []*Param, resultList []*Result) {
 	c.Results = resultList
 
 	dg.Ctors = append(dg.Ctors, c)
+	dg.ctorMap[c.ID] = c
+}
+
+func (dg *Graph) failNode(r *Result, pof bool) {
+	if pof {
+		dg.addRootCause(r)
+	} else {
+		dg.addTransitiveFailure(r)
+	}
+}
+
+// MissingNodes adds missing nodes to the list of failed Results in the graph.
+func (dg *Graph) MissingNodes(results []*Result) {
+	pof := len(dg.Failed.RootCauses) == 0
+
+	for _, r := range results {
+		dg.failNode(r, pof)
+	}
+}
+
+// FailNodes adds results to the list of failed Results in the graph, and
+// updates the state of the constructor with the given id accordingly.
+func (dg *Graph) FailNodes(results []*Result, id CtorID) {
+	pof := len(dg.Failed.RootCauses) == 0
```

Comment:
When is `len(dg.Failed.RootCauses) != 0`? It seems like once a failure is encountered, there will always be at least one root (that therefore create transitive failures). If this is the case, after the initial root failure is added, this will always be > 0, and everything would be colored as red since `pof` is true. 

I'm assuming that transitive failures are appropriately colored, however, so what piece am I misunderstanding?

Reply:
`pof` is true when there are no existing root causes. Once the first failed node(s) are recorded, they are added to `dg.Failed.RootCauses`, and all other errors will have `pof = false` and be added as transitive failures.

Reply:
Can we add a quick comment about this in the places where we're doing `pof := `

    // This failure is the root cause if there are no other failures.
    pof := len(...) == 0

Reply:
+1 makes sense. The comment would be helpful.

Reply:
I added comments, also renamed `pof` to `isRootCause`.